aptos-move/aptos-vm/src/move_vm_ext/aggregator_extension.rs

```diff
@@ -0,0 +1,644 @@
+            for (key, aggregator) in table {
+                let Aggregator {
+                    state,
+                    value,
+                    limit,
+                    ..
+                } = aggregator;
+
+                let op = match state {
+                    AggregatorState::Data => AggregatorChange::Write(value),
+                    AggregatorState::PositiveDelta => {
+                        let delta_op = DeltaOp::Addition { value, limit };
+                        AggregatorChange::Delta(delta_op)
+                    }
+                };
+                change.insert(key, Some(op));
+            }
+
+            // Add changes that delete aggregators.
+            for key in removed_aggregators {
+                change.insert(key, None);
+            }
+
+            changes.insert(table_handle, change);
+        }
+
+        AggregatorChangeSet { changes }
+    }
+}
+
+// ================================= Natives =================================
+
+/// All aggregator native functions. For more details, refer to code in
+/// `aggregator_table.move`.
+pub fn aggregator_natives(aggregator_addr: AccountAddress) -> NativeFunctionTable {
+    native_functions::make_table(
+        aggregator_addr,
+        &[
+            ("aggregator", "add", native_add),
+            ("aggregator", "read", native_read),
+            ("aggregator", "remove_aggregator", native_remove_aggregator),
+            ("aggregator", "sub", native_sub),
+            ("aggregator_table", "new_aggregator", native_new_aggregator),
+        ],
+    )
+}
+
+fn native_new_aggregator(
+    context: &mut NativeContext,
+    _ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> PartialVMResult<NativeResult> {
+    assert!(args.len() == 2);
+
+    // Extract fields: `limit` of the new aggregator and a `table_handle`
+    // associated with the associated aggregator table.
+    let limit = pop_arg!(args, IntegerValue).value_as::<u128>()?;
+    let table_handle = get_table_handle(&pop_arg!(args, StructRef))?;
+
+    // Get the current aggregator table.
+    let aggregator_context = context.extensions().get::<NativeAggregatorContext>();
+    let mut aggregator_table_data = aggregator_context.aggregator_table_data.borrow_mut();
+    let aggregator_table = aggregator_table_data.get_or_create_aggregator_table(table_handle);
+
+    // Every aggregator instance has a unique key. Here we can reuse the
+    // strategy from `table` implementation: taking hash of transaction and
+    // number of aggregator instances created so far and truncating them to
+    // 128 bits.
+    let txn_hash_buffer = serialize(&aggregator_context.txn_hash);
+    let num_aggregators_buffer = serialize(&aggregator_table.num_aggregators());
+
+    let mut sha3 = Sha3::v256();
+    sha3.update(&txn_hash_buffer);
+    sha3.update(&num_aggregators_buffer);
+
+    let mut key_bytes = [0_u8; 16];
+    sha3.finalize(&mut key_bytes);
+    let key = deserialize(&key_bytes.to_vec());
```

Comment:
Looks kind of strange that you deserialize from a sha256 hash.

Reply:
My intention was to construct `u128` from bytes, but this can also be done with `to_be_bytes()`